src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/Inputs/SARIF1008.PhysicalLocationPropertiesMustBeConsistent_Invalid.sarif

```diff
@@ -28,6 +28,154 @@
+                  "endLine": 10
+                },
+                "contextRegion": {
+                  "startLine": 2,
+                  "endLine": 4
+                }
+              }
+            }
+          ]
+        },
+        {
+          "ruleId": "TST0004",
+          "level": "error",
+          "message": {
+            "text": "Bad result - fully overlapping region and contextregion (line/column based)."
+          },
+          "locations": [
+            {
+              "physicalLocation": {
+                "artifactLocation": {
+                  "uri": "src/test.c"
+                },
+                "region": {
+                  "startLine": 2,
+                  "endLine": 4
+                },
+                "contextRegion": {
+                  "startLine": 2,
+                  "endLine": 4
+                }
+              }
+            }
+          ]
+        },
+        {
+          "ruleId": "TST0005",
+          "level": "error",
+          "message": {
+            "text": "Bad result - text contextregion smaller than region (offset based)."
+          },
+          "locations": [
+            {
+              "physicalLocation": {
+                "artifactLocation": {
+                  "uri": "src/test.c"
+                },
+                "region": {
+                  "charOffset": 15,
+                  "charLength": 30
+                },
+                "contextRegion": {
+                  "charOffset": 15,
+                  "charLength": 20
+                }
+              }
+            }
+          ]
+        },
+        {
+          "ruleId": "TST0006",
+          "level": "error",
+          "message": {
+            "text": "Bad result - non overlapping text region and contextregion (offset based)."
+          },
+          "locations": [
+            {
+              "physicalLocation": {
+                "artifactLocation": {
+                  "uri": "src/test.c"
+                },
+                "region": {
+                  "charOffset": 1,
+                  "charLength": 10
+                },
+                "contextRegion": {
+                  "charOffset": 15,
+                  "charLength": 20
+                }
+              }
+            }
+          ]
+        },
+        {
+          "ruleId": "TST0007",
+          "level": "error",
+          "message": {
+            "text": "Bad result - fully overlapping binary region and contextregion (offset based)."
+          },
+          "locations": [
+            {
+              "physicalLocation": {
+                "artifactLocation": {
+                  "uri": "src/test.c"
+                },
+                "region": {
+                  "byteOffset": 10,
+                  "byteLength": 100
+                },
+                "contextRegion": {
+                  "byteOffset": 10,
+                  "byteLength": 100
+                }
+              }
+            }
+          ]
         }
       ],
```

Comment:
>] [](start = 6, length = 1)

We also need tests that check the defaults (like, if endLine is absent, it defaults to startLine, and if endColumn is absent, it defaults to "end of line"). There is a helper somewhere in the SDK that fills out the region properties. This might be as simple as calling that helper before you do your logic.

Reply:
lets chat offline about this quickly, will ping u

---
In reply to: [443870943](https://github.com/microsoft/sarif-sdk/pull/1925#discussion_r443870943) [](ancestors = 443870943)

Reply:
The idea is that the region properties have defaults that will throw off your calculations. Your calculations assume that all the properties are present. But for example consider the default for `endLine`. If it's missing it defaults to `startLine`. So here:

```
"region": {
    "startLine": 5  # endLine defaults to 5, but in the code you will see region.EndLine = -1.
},
"contextRegion": {
    "startLine": 5,
    "endLine": 5    # In the code, region.EndLine = 5.
}
```

So you will calculate the they have the same startLine, but contextRegion.EndLine > region.EndLine because 5 > -1. So you'll decide that contextRegion is a proper superset of region, whereas in fact they are identical.

Similarly for `endColumn`. If `startLine` and `endLine` are the same but `endColumn` is absent, it defaults to "one past the end of the line". 

```
"region": {
    "startLine": 2,
    "startColumn": 5   # endLine defaults to startLine, so endColumn defaults to one past end of line. But in the code you'll see region.EndColumn = -1 because it's absent.
},
"contextRegion": {
    "startLine": 2,
    "startColumn": 5,
    "endColumn": 6,
}
```

So again you will thing contextRegion is a proper superset because 6 > -1, but in fact it is a subset.

I thought there was a helper that filled out the defaults for a region object so your calculations will work. I'll ask Michael.

---
In reply to: [443893752](https://github.com/microsoft/sarif-sdk/pull/1925#discussion_r443893752) [](ancestors = 443893752,443870943)

Reply:
great tests, i have added these in now (and the corresponding code)

---
In reply to: [444253628](https://github.com/microsoft/sarif-sdk/pull/1925#discussion_r444253628) [](ancestors = 444253628,443893752,443870943)